### PR TITLE
feat(STONEINTG-483): add requeueDelay to ensure* functions

### PR DESCRIPTION
**Pull request by hongweiliu17**
Read the code about `RequeueRequest`, `CancelRequest` and `error` in [ReconcileHandler(operations []ReconcileOperation)](https://github.com/redhat-appstudio/operator-goodies/blob/main/reconciler/handler.go#L23-L35), according to the definitions of [RequeueWithError(error)](https://github.com/redhat-appstudio/operator-goodies/blob/main/reconciler/results.go#L45) and [RequeueOnErrorOrStop(error)](https://github.com/redhat-appstudio/operator-goodies/blob/main/reconciler/results.go#L75), these two functions are the same when the `error` is not `nil`, that is to `requeue`.

1. Replace `RequeueOnErrorOrStop(error)` with `RequeueWithError(error)` when error is not `nil` to be consistent in all  adapters
2. Call `RequeueWithError(err)` when failing to update scenario to requeue

Signed-off-by: Hongwei Liu <hongliu@redhat.com>

### CODE REVIEW

File: controllers/pipeline/pipeline_adapter.go [hunk 1]
```diff
@@ -36,6 +36,10 @@ import (
 	"sigs.k8s.io/controller-runtime/pkg/client"
 )
 
+const (
+	requeueDelay = 10 * time.Second
```
[the remark this answers is not in the document]
Comment (hongweiliu17) [reply controllers/pipeline/pipeline_adapter.go:40]: I reference the time in https://github.com/redhat-appstudio/managed-gitops/tree/main/appstudio-controller/controllers/appstudio.redhat.com when they meet some intermittent error, but I'd like to know your ideas. Thanks.

File: controllers/snapshot/snapshot_adapter.go [hunk 2]
```diff
@@ -251,7 +256,7 @@ TestScenarioLoop:
 				"snapshot", a.snapshot.Name,
 				"environment.Name", copyEnv.Name,
 				"snapshot.Spec.Components", a.snapshot.Spec.Components)
-			return reconciler.RequeueWithError(err)
```
Comment (MartinBasti) [inline controllers/snapshot/snapshot_adapter.go:254]: Why this is not error anymore?

File: controllers/snapshot/snapshot_adapter.go [hunk 3]
```diff
@@ -237,7 +242,7 @@ TestScenarioLoop:
 
 		if err != nil {
 			a.logger.Error(err, "Copying of environment failed")
-			return reconciler.RequeueWithError(err)
+			return reconciler.RequeueAfter(requeueDelay, err)
```
Comment (MartinBasti) [inline controllers/snapshot/snapshot_adapter.go:245]: Why do we need manual requeueDelay, when we have by default exponential backoff retries when error happen in RequeueWithError.

**Review by MartinBasti — CHANGES_REQUESTED** [review]
I don't understand purpose of these changes, I may need explanation like for 5y old

EDIT: mean point 2. RequeueAfter changes

File: controllers/snapshot/snapshot_adapter.go (hunk 2 above)
Comment (hongweiliu17) [reply controllers/snapshot/snapshot_adapter.go:254]: Comparing with `RequeueWithError`, `reconciler.RequeueAfter(requeueDelay, err)` is having an additional delay with requeue.

**Comment by hongweiliu17** [comment]
> I don't understand purpose of these changes, I may need explanation like for 5y old
> 
> EDIT: mean point 2. RequeueAfter changes

I'll prepare a short demo for it.

**Comment by hongweiliu17** [comment]
> I don't understand purpose of these changes, I may need explanation like for 5y old
> 
> EDIT: mean point 2. RequeueAfter changes

I make a [demo](https://drive.google.com/file/d/1I4W-0UiJ1361Axaw_iafXv_9NI-leJLY/view) for this PR.

File: controllers/snapshot/snapshot_controller.go [hunk 4]
```diff
@@ -156,5 +159,8 @@ func setupControllerWithManager(manager ctrl.Manager, reconciler *Reconciler) er
 		For(&applicationapiv1alpha1.Snapshot{}).
 		WithEventFilter(predicate.Or(
 			gitops.IntegrationSnapshotChangePredicate())).
+		WithOptions(controller.Options{
+			RateLimiter: workqueue.NewItemExponentialFailureRateLimiter(2*time.Second, 1*time.Hour),
```
Comment (MartinBasti) [inline controllers/snapshot/snapshot_controller.go:163]: Why this change? Was default rate limit 2-100sec not enough?

Comment (hongweiliu17) [reply controllers/snapshot/snapshot_controller.go:163]: I wanted to find the place to set the default rate limit but I didn't find it. Can you help to paste the link. I want to read it. Thanks!

**Comment by MartinBasti** [comment]
According this article https://danielmangum.com/posts/controller-runtime-client-go-rate-limiting/

image

Is the RequeueAfter value even used when `err != nil`?

File: controllers/snapshot/snapshot_controller.go (hunk 4 above)
Comment (MartinBasti) [reply controllers/snapshot/snapshot_controller.go:163]: https://github.com/kubernetes/client-go/blob/master/util/workqueue/default_rate_limiters.go#L39 This one

I had a typo in my previous msg, 1000sec (~17min what we see in logs)

Comment (hongweiliu17) [reply controllers/snapshot/snapshot_controller.go:163]: Thanks very much. Let me remove it.

**Comment by hongweiliu17** [comment]
> 
If the `err == nil` and  `requeue is false`, don't requeue, otherwise, requeue.

**Comment by hongweiliu17** [comment]
> According this article https://danielmangum.com/posts/controller-runtime-client-go-rate-limiting/
> 
> image
> 
> Is the RequeueAfter value even used when `err != nil`?

If the `err == nil` and  `requeue is false`, don't requeue, otherwise, requeue. `RequeueAfter` is defined from `RequeueAfter: result.RequeueDelay`, otherwise, it is set to `0`, I think.

**Comment by MartinBasti** [comment]
> > According this article https://danielmangum.com/posts/controller-runtime-client-go-rate-limiting/
> > image
> > Is the RequeueAfter value even used when `err != nil`?
> 
> If the `err == nil` and `requeue is false`, don't requeue, otherwise, requeue. `RequeueAfter` is defined from `RequeueAfter: result.RequeueDelay`, otherwise, it is set to `0`, I think.

Here, err is not nil, but we have reqeueuDelay. According the table it means, that requeueDelay will be ignored and just standard error rate limiting will be used.
```
		if err != nil {
			a.logger.Error(err, "Copying of environment failed")
			return reconciler.RequeueAfter(requeueDelay, err)
```

**Comment by hongweiliu17** [comment]
> > > According this article https://danielmangum.com/posts/controller-runtime-client-go-rate-limiting/
> > > image
> > > Is the RequeueAfter value even used when `err != nil`?
> > 
> > 
> > If the `err == nil` and `requeue is false`, don't requeue, otherwise, requeue. `RequeueAfter` is defined from `RequeueAfter: result.RequeueDelay`, otherwise, it is set to `0`, I think.
> 
> Here, err is not nil, but we have reqeueuDelay. According the table it means, that requeueDelay will be ignored and just standard error rate limiting will be used.
> 
> ```
> 		if err != nil {
> 			a.logger.Error(err, "Copying of environment failed")
> 			return reconciler.RequeueAfter(requeueDelay, err)
> ```

Yes. According to the logic in [controller-runtime](https://github.com/kubernetes-sigs/controller-runtime/blob/16bf3ad036b908d897543c415fcc0bafc5cec711/pkg/internal/controller/controller.go#L297-L311), https://github.com/redhat-appstudio/operator-goodies/blob/main/reconciler/handler.go#L27 looks incorrect, the `RequeueAfter` in code `return ctrl.Result{RequeueAfter: result.RequeueDelay}, err` can't take effect totally.

**Comment by MartinBasti** [comment]
> > > > According this article https://danielmangum.com/posts/controller-runtime-client-go-rate-limiting/
> > > > image
> > > > Is the RequeueAfter value even used when `err != nil`?
> > > 
> > > 
> > > If the `err == nil` and `requeue is false`, don't requeue, otherwise, requeue. `RequeueAfter` is defined from `RequeueAfter: result.RequeueDelay`, otherwise, it is set to `0`, I think.
> > 
> > 
> > Here, err is not nil, but we have reqeueuDelay. According the table it means, that requeueDelay will be ignored and just standard error rate limiting will be used.
> > ```
> > 		if err != nil {
> > 			a.logger.Error(err, "Copying of environment failed")
> > 			return reconciler.RequeueAfter(requeueDelay, err)
> > ```
> 
> Yes. According to the logic in [controller-runtime](https://github.com/kubernetes-sigs/controller-runtime/blob/16bf3ad036b908d897543c415fcc0bafc5cec711/pkg/internal/controller/controller.go#L297-L311), https://github.com/redhat-appstudio/operator-goodies/blob/main/reconciler/handler.go#L27 looks incorrect, the `RequeueAfter` in code `return ctrl.Result{RequeueAfter: result.RequeueDelay}, err` can't take effect totally.

Ah nice, thanks for confirming. So we can remove `RequeueDelay` change from this PR and open an issue for them (for operator-goodies).

**Comment by hongweiliu17** [comment]
> > > > > According this article https://danielmangum.com/posts/controller-runtime-client-go-rate-limiting/
> > > > > image
> > > > > Is the RequeueAfter value even used when `err != nil`?
> > > > 
> > > > 
> > > > If the `err == nil` and `requeue is false`, don't requeue, otherwise, requeue. `RequeueAfter` is defined from `RequeueAfter: result.RequeueDelay`, otherwise, it is set to `0`, I think.
> > > 
> > > 
> > > Here, err is not nil, but we have reqeueuDelay. According the table it means, that requeueDelay will be ignored and just standard error rate limiting will be used.
> > > ```
> > > 		if err != nil {
> > > 			a.logger.Error(err, "Copying of environment failed")
> > > 			return reconciler.RequeueAfter(requeueDelay, err)
> > > ```
> > 
> > 
> > Yes. According to the logic in [controller-runtime](https://github.com/kubernetes-sigs/controller-runtime/blob/16bf3ad036b908d897543c415fcc0bafc5cec711/pkg/internal/controller/controller.go#L297-L311), https://github.com/redhat-appstudio/operator-goodies/blob/main/reconciler/handler.go#L27 looks incorrect, the `RequeueAfter` in code `return ctrl.Result{RequeueAfter: result.RequeueDelay}, err` can't take effect totally.
> 
> Ah nice, thanks for confirming. So we can remove `RequeueDelay` change from this PR and open an issue for them (for operator-goodies).

Do you know the place we can file issue for it? cc @dirgim 

Maybe we can just set a requeueAfter in our code now, then we don't need to update our code once operator-goodies fix it?

File: controllers/pipeline/pipeline_adapter.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (hongweiliu17) [reply controllers/pipeline/pipeline_adapter.go]: Removed the manual delay since we can't implement it currently due to the issue in https://github.com/redhat-appstudio/operator-goodies/blob/main/reconciler/handler.go#L27.

File: controllers/snapshot/snapshot_adapter.go (hunk 2 above)
Comment (hongweiliu17) [reply controllers/snapshot/snapshot_adapter.go]: Removed the manual delay since we can't implement it currently due to the issue in https://github.com/redhat-appstudio/operator-goodies/blob/main/reconciler/handler.go#L27.

File: controllers/snapshot/snapshot_adapter.go (hunk 3 above)
Comment (hongweiliu17) [reply controllers/snapshot/snapshot_adapter.go]: Removed the manual delay since we can't implement it currently due to the issue in https://github.com/redhat-appstudio/operator-goodies/blob/main/reconciler/handler.go#L27.

File: controllers/snapshot/snapshot_controller.go (hunk 4 above)
Comment (hongweiliu17) [reply controllers/snapshot/snapshot_controller.go]: Removed.

**Comment by hongweiliu17** [comment]
> > > > > > According this article https://danielmangum.com/posts/controller-runtime-client-go-rate-limiting/
> > > > > > image
> > > > > > Is the RequeueAfter value even used when `err != nil`?
> > > > > 
> > > > > 
> > > > > If the `err == nil` and `requeue is false`, don't requeue, otherwise, requeue. `RequeueAfter` is defined from `RequeueAfter: result.RequeueDelay`, otherwise, it is set to `0`, I think.
> > > > 
> > > > 
> > > > Here, err is not nil, but we have reqeueuDelay. According the table it means, that requeueDelay will be ignored and just standard error rate limiting will be used.
> > > > ```
> > > > 		if err != nil {
> > > > 			a.logger.Error(err, "Copying of environment failed")
> > > > 			return reconciler.RequeueAfter(requeueDelay, err)
> > > > ```
> > > 
> > > 
> > > Yes. According to the logic in [controller-runtime](https://github.com/kubernetes-sigs/controller-runtime/blob/16bf3ad036b908d897543c415fcc0bafc5cec711/pkg/internal/controller/controller.go#L297-L311), https://github.com/redhat-appstudio/operator-goodies/blob/main/reconciler/handler.go#L27 looks incorrect, the `RequeueAfter` in code `return ctrl.Result{RequeueAfter: result.RequeueDelay}, err` can't take effect totally.
> > 
> > 
> > Ah nice, thanks for confirming. So we can remove `RequeueDelay` change from this PR and open an issue for them (for operator-goodies).
> 
> Do you know the place we can file issue for it? cc @dirgim
> 
> Maybe we can just set a requeueAfter in our code now, then we don't need to update our code once operator-goodies fix it?

Removed the manual delay since we can't implement it currently due to the issue in https://github.com/redhat-appstudio/operator-goodies/blob/main/reconciler/handler.go#L27.

Thanks for the review! @MartinBasti

**Comment by MartinBasti** [comment]
> Maybe we can just set a requeueAfter in our code now, then we don't need to update our code once operator-goodies fix it?

I don't understand how goodies can fix it, this is how k8s works, if there is an error `requeueAfter` won't apply at all

**Review by MartinBasti — APPROVED** [review]

**Review by jinqi7 — APPROVED** [review]
LGTM
